File: farmconnect-backend/product-service/src/main/java/com/farmconnect/productservice/service/ProductService.java
package com.farmconnect.productservice.service;

import org.springframework.beans.factory.annotation.Autowired;
import com.farmconnect.productservice.dto.ProductRequest;
import com.farmconnect.productservice.dto.ProductResponse;
import com.farmconnect.productservice.model.Product;
import com.farmconnect.productservice.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private FileStorageService fileStorageService;

    public ProductResponse updateProductImage(Long productId, String imageUrl) {
    Product product = productRepository.findById(productId)
            .orElseThrow(() -> new RuntimeException("Product not found!"));
    
    // Delete old image if exists
    if (product.getImageUrl() != null && !product.getImageUrl().isEmpty()) {
        String oldFileName = extractFileNameFromUrl(product.getImageUrl());
        if (oldFileName != null) {
            fileStorageService.deleteFile(oldFileName);
        }
    }
    
        product.setImageUrl(imageUrl);
        Product updatedProduct = productRepository.save(product);
        return new ProductResponse(updatedProduct);
    }

    // Helper method to extract filename from URL
    private String extractFileNameFromUrl(String url) {
        if (url == null || url.isEmpty()) {
            return null;
        }
        int lastSlashIndex = url.lastIndexOf('/');
        if (lastSlashIndex != -1 && lastSlashIndex < url.length() - 1) {
            return url.substring(lastSlashIndex + 1);
        }
        return null;
    }
    
    public ProductResponse createProduct(ProductRequest productRequest) {
        Product product = new Product();
        product.setFarmerId(productRequest.getFarmerId());
        product.setProductName(productRequest.getProductName());
        product.setDescription(productRequest.getDescription());
        product.setPrice(productRequest.getPrice());
        product.setQuantityAvailable(productRequest.getQuantityAvailable());
        product.setCategory(productRequest.getCategory());
        product.setImageUrl(productRequest.getImageUrl());
        product.setIsAvailable(productRequest.getIsAvailable());

        Product savedProduct = productRepository.save(product);
        return new ProductResponse(savedProduct);
    }

    public List<ProductResponse> getAllProducts() {
        return productRepository.findAll().stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());
    }

    public List<ProductResponse> getAvailableProducts() {
        return productRepository.findByIsAvailableTrue().stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());
    }

    public ProductResponse getProductById(Long productId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found!"));
        return new ProductResponse(product);
    }

    public List<ProductResponse> getProductsByFarmer(Long farmerId) {
        return productRepository.findByFarmerId(farmerId).stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());
    }

    public List<ProductResponse> searchProducts(String keyword) {
        return productRepository.findByProductNameContainingIgnoreCase(keyword).stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());
    }

    public List<ProductResponse> getProductsByCategory(String category) {
        return productRepository.findByCategory(category).stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());
    }

    public ProductResponse updateProduct(Long productId, ProductRequest productRequest) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found!"));

        product.setProductName(productRequest.getProductName());
        product.setDescription(productRequest.getDescription());
        product.setPrice(productRequest.getPrice());
        product.setQuantityAvailable(productRequest.getQuantityAvailable());
        product.setCategory(productRequest.getCategory());
        product.setImageUrl(productRequest.getImageUrl());
        product.setIsAvailable(productRequest.getIsAvailable());

        Product updatedProduct = productRepository.save(product);
        return new ProductResponse(updatedProduct);
    }

    public void deleteProduct(Long productId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found!"));
        productRepository.delete(product);
    }

    public void updateProductQuantity(Long productId, Integer quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found!"));
        product.setQuantityAvailable(quantity);
        productRepository.save(product);
    }
}

File: farmconnect-frontend/src/services/productService.js
import axios from 'axios';

const API_URL = 'http://localhost:8082/api/products';

const productService = {
  // Create new product
  createProduct: async (productData, token) => {
    const response = await axios.post(API_URL, productData, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  },

  // Get all products
  getAllProducts: async () => {
    const response = await axios.get(API_URL);
    return response.data;
  },

  // Get available products
  getAvailableProducts: async () => {
    const response = await axios.get(`${API_URL}/available`);
    return response.data;
  },

  // Get product by ID
  getProductById: async (productId) => {
    const response = await axios.get(`${API_URL}/${productId}`);
    return response.data;
  },

  // Get products by farmer
  getProductsByFarmer: async (farmerId) => {
    const response = await axios.get(`${API_URL}/farmer/${farmerId}`);
    return response.data;
  },

  // Search products
  searchProducts: async (keyword) => {
    const response = await axios.get(`${API_URL}/search?keyword=${keyword}`);
    return response.data;
  },

  // Get products by category
  getProductsByCategory: async (category) => {
    const response = await axios.get(`${API_URL}/category/${category}`);
    return response.data;
  },

  // Update product
  updateProduct: async (productId, productData, token) => {
    const response = await axios.put(`${API_URL}/${productId}`, productData, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  },

  // Delete product
  deleteProduct: async (productId, token) => {
    const response = await axios.delete(`${API_URL}/${productId}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  },

  // Update product quantity
  updateProductQuantity: async (productId, quantity, token) => {
    const response = await axios.put(
      `${API_URL}/${productId}/quantity?quantity=${quantity}`,
      {},
      { headers: { Authorization: `Bearer ${token}` } }
    );
    return response.data;
  },
};

export default productService;